src/test/java/cat/udl/eps/softarch/demo/steps/CreateDonationStepDefs.java

```diff
@@ -26,134 +24,47 @@ public class CreateDonationStepDefs {
     @Autowired
     private StepDefs stepDefs;
     @Autowired
-    private UserRepository userRepository;
-    @Autowired
     private TakeRepository takeRepository;
     @Autowired
     private DonationRepository donationRepository;
 
-    @And("There is a valid Donor")
-    public void thereIsAValidDonor() {
-        Donor donor = createValidDonor();
-        userRepository.save(donor);
-        Assert.assertNotEquals(0, userRepository.count());
+    @Autowired
+    private DonorRepository donorRepository;
+
+    private Take take;
+
+    private Donor donor;
+
+
+    @And("User {string} is the donor")
+    public void thereIsAValidDonor(String name) {
+        this.donor = donorRepository.findById(name).orElseThrow();
     }
 
-    @And("There is a valid Take")
+    @And("A valid take action exists")
     public void thereIsAValidTake() {
-        Take take = createValidTake();
-        takeRepository.save(take);
-        Assert.assertNotEquals(0, takeRepository.count());
+        this.take = TakeMother.getValidTake();
+        takeRepository.save(this.take);
     }
 
-    @When("I create a new valid donation")
+    @When("The donor creates a donation from the take action")
     public void createDonation() throws Exception {
-        Donation donation = createValidDonation();
-
-        stepDefs.result = stepDefs.mockMvc.perform(
-                post("/donations")
-                        .contentType(MediaType.APPLICATION_JSON)
-                        .content(stepDefs.mapper.writeValueAsString(donation))
-                        .accept(MediaType.APPLICATION_JSON)
-                        .with(AuthenticationStepDefs.authenticate()))
-                .andDo(print());
+        var donation = DonationMother.getValidDonationFor(donor, take);
+        stepDefs.result = stepDefs.mockMvc.perform(post("/donations").contentType(MediaType.APPLICATION_JSON_VALUE).characterEncoding("utf-8").content(stepDefs.mapper.writeValueAsString(donation)).accept(MediaType.APPLICATION_JSON).with(AuthenticationStepDefs.authenticate())).andDo(print());
     }
 
     @But("There is no Donor")
     public void thereIsNoValidDonor() {
-        for (User user : userRepository.findAll()) {
-            Assert.assertFalse(user instanceof Donor);
-        }
+        assertNull(donor);
     }
 
     @But("There is no Take")
     public void thereIsNoValidTake() {
-        assert takeRepository.count() == 0;
+        assertNull(take);
```

Comment:
With this implementation, you are just checking the local test variable but not the application state. Better to use the previous implementation, checking though the repository that the databases doesn't include any Take or Donor

Reply:
Done 😃